### PR TITLE
[Clang] CWG2789 Overload resolution with implicit and explicit object…

### PR DESCRIPTION
**Pull request by cor3ntin**
… member functions

Implement the resolution to CWG2789 from
https://wiki.edg.com/pub/Wg21kona2023/StrawPolls/p3046r0.html

The DR page is not updated because the issue has not made it to a published list yet.

### CODE REVIEW

File: clang/test/CXX/drs/dr27xx.cpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    constexpr void f(this S&) requires true{}; // #2
+
+    using Base<T>::g;
+    using Base2<T>::g;
+};
+
+void test() {
+    S<> s;
+    s.f();
+    s.g(); // expected-error {{call to member function 'g' is ambiguous}}
```
Comment (Endilll) [inline clang/test/CXX/drs/dr27xx.cpp:28]: I believe our current best practices on handling errors with notes in DR tests are showcased here: https://github.com/llvm/llvm-project/blob/fe3c4218438adeebba4914a57b4d382e3a947394/clang/test/CXX/drs/dr13xx.cpp#L505
Notes being scattered across the test makes it harder to match the test against compiler output on e.g. Compiler Explorer.

**Review by Endilll — CHANGES_REQUESTED** [review]

File: clang/lib/Sema/SemaOverload.cpp [hunk 2]
```diff
@@ -10121,22 +10143,41 @@ static bool haveSameParameterTypes(ASTContext &Context, const FunctionDecl *F1,
...
-        S.FunctionParamTypesAreEqual(PT1, PT2, nullptr,
-                                     Cand1.isReversed() ^ Cand2.isReversed()))
-      return true;
+                                           const OverloadCandidate &Cand1,
+                                           const OverloadCandidate &Cand2) {
+  if (!Cand1.Function || !Cand2.Function)
+    return false;
+
+  auto *Fn1 = Cand1.Function;
+  auto *Fn2 = Cand2.Function;
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaOverload.cpp:10156]: Please spell out the types explicitly as they're not spelled by the initializer.

**Review by AaronBallman — APPROVED** [review]
Changes LGTM modulo some comments on the review.

**Comment by cor3ntin** [comment]
@Endilll I addressed your feedback!

**Review by Endilll — APPROVED** [review]

**Pull request merged**